### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by O-S-K** [gilzoide/unity-update-manager#20]
I tried it and it seems Managed Update runs slower and wastes more garbage, I haven't tried it with jobs and I want to ask you if I implemented it wrong

![image](https://github.com/user-attachments/assets/7e886d33-ca0c-4414-a9df-a780fe4cec7f)

**Comment by gilzoide** [comment]
Hey @O-S-K.

Regarding bigger memory usage:
1. The Update Manager maintains some [lists](https://github.com/gilzoide/unity-update-manager/blob/main/Runtime/UpdateManager.cs#L65-L67) for managing updatable objects.
2. There's also a [static Dictionary in the ProfileMarkerMap](https://github.com/gilzoide/unity-update-manager/blob/main/Runtime/Internal/ProfilerMarkerMap.cs#L25), which only exists in the Unity editor and development builds. You can disable profile markers using the scripting symbol `UPDATE_MANAGER_DISABLE_PROFILER_MARKERS`, if you want.

As for the difference in runtime, profile markers are likely the culprit. You can disable them by adding the `UPDATE_MANAGER_DISABLE_PROFILER_MARKERS` scripting symbol to your project.
There are performance tests in the package, you can run them using Unity's [Test Runner](https://docs.unity3d.com/Manual/testing-editortestsrunner.html). Here are the results from a run in my machine with profile markers enabled:
<img width="849" alt="Screenshot 2024-11-30 at 11 01 41" src="https://github.com/user-attachments/assets/6db386bb-fd9d-4687-b09b-02b4cb036c73">
Versus the results in the same machine, with profile markers disabled:
<img width="849" alt="Screenshot 2024-11-30 at 11 00 19" src="https://github.com/user-attachments/assets/19c6804f-256b-4758-9000-b1a0b62fa6e6">
Notice how the median and standard deviation are higher with managed updates than plain update when profile markers are enabled, but lower when profile markers are disabled.

Maybe it makes sense to keep them disabled by default and document how to enable, rather than the other way around. It would likely be less confusing. What do you think?

<hr/>

By the way, how many objects are being updated? The Update Manager pattern works best when you have hundreds or thousands of objects to update. If there's only one, the overhead is definitely not worth using it.

**Comment by O-S-K** [comment]
In theory

UpdateManager: There is usually only one call to Update in the game loop, then the logic is managed via lists or delegates. This reduces the overhead of Unity Message Dispatch.

Update 1000: Unity has to call each Update function on each GameObject, which results in additional overhead from internal message distribution.

Is that correct and like yours when to use JobUpdate and when to use UpdateManager, are there any considerations when using them, thanks for your reply

**Comment by gilzoide** [comment]
> UpdateManager: There is usually only one call to Update in the game loop, then the logic is managed via lists or delegates. This reduces the overhead of Unity Message Dispatch.

> Update 1000: Unity has to call each Update function on each GameObject, which results in additional overhead from internal message distribution.

Yes, that is correct. The idea is exactly to remove the overhead from the C++ <-> C# interop that occurs when Unity calls Update/LateUpdate/FixedUpdate methods. For 1000 objects, using the Update Manager you get only 1 interop call and 1000 C# calls, which is faster than having 1000 interop calls.

> when to use JobUpdate and when to use UpdateManager

Using UpdateManager is much simpler, as it is just like the usual Update, but made through the manager. You just have to inherit `AManagedBehaviour, IUpdatable` instead of `MonoBehaviour` and have the `ManagedUpdate` function instead of the regular `Update`, just like you did in your tests.

Job update managers, on the other hand, use Unity's [C# Job System](https://docs.unity3d.com/Manual/job-system-overview.html). Using the Job system is much more complex and has a lot more restrictions.

Pros:
- Jobs run in background threads, so you free Unity's main thread to do other work
- Jobs may run in parallel, speeding the whole thing up
- Jobs may be compiled to highly optimized native code using the [Burst](https://docs.unity3d.com/Packages/com.unity.burst@1.8/manual/index.html) compiler, speeding the whole thing even more
- Unity manages all multithreading synchronization for you, all you need to do is create jobs correctly, schedule them, and wait for them to complete if necessary

Cons:
- Jobs only support `struct` types. You can't have any reference types on then, not even `string` or `List<>`. You can use Unity's [Native Collections](https://docs.unity3d.com/Packages/com.unity.collections@2.5/manual/index.html) package for these cases, but it requires more code and much more care to avoid crashes and memory leaks.
- Jobs run on background threads. As most of the Unity API is not thread safe and requires being called from the main thread, code that run on jobs is more limited on what they can accomplish.

My suggestion is: just stick to UpdateManager for most things and use JobUpdateManager only if you really understand how to use the C# Job System and really know what you're doing.

**Comment by O-S-K** [comment]
my friend i want to hug you, thanks for sharing so enthusiastically

**Comment by gilzoide** [comment]
No problem, always glad to help ^^

**Comment by O-S-K** [comment]
I have a question for you, if in my project there are 10-20 updates, but in the same frame only 2-3 updates are enabled and executed, is there too much difference in performance?

**Comment by gilzoide** [comment]
> is there too much difference in performance?

The only way to know for sure is profiling. Always profile before optimizing something.

The benefits of the Update Manager pattern scales with the quantity of objects being processed. I guess only 2-3 updates will make no noticeable difference, even on lower end platforms. Again, this is a guess and the only way to know for sure is actually profiling on the target device.

Just to remind you, you don't need to use Update Manager on a single script. You can use it in several scripts and all of them will be updated together. Also as a reminder, beware that the UpdateManager doesn't preserve the same [Script Execution Order](https://docs.unity3d.com/Manual/class-MonoManager.html), so you can't rely on update order when using it.

**Comment by gilzoide** [comment]
I'm closing this issue, since it has been answered. Feel free to open a new Issue or Discussion if necessary.
